### PR TITLE
feat: add OpenTelemetry tracing with build and package spans

### PR DESCRIPTION
**Pull request by corneliusludmann**
## What

Add OpenTelemetry infrastructure for distributed tracing with support for W3C Trace Context propagation from CI systems.

## Why

Provides visibility into build performance and behavior, enabling:
- Build lifecycle tracking (start to finish)
- Individual package build timing
- Build phase durations (prep, pull, lint, test, build, package)
- Cache hit/miss metrics
- GitHub Actions context capture
- Distributed tracing across CI pipelines

### Core Implementation
- **Telemetry Package** (`pkg/leeway/telemetry/`):
  - `InitTracer()` - Initializes OTLP HTTP exporter
  - `Shutdown()` - Flushes spans with 5-second timeout
  - `ParseTraceContext()` - Parses W3C Trace Context headers
  - `FormatTraceContext()` - Formats span context to W3C format
  - `ValidateTraceParent()` - Validates traceparent format

- **OTelReporter** (`pkg/leeway/reporter.go`):
  - Implements `Reporter` interface
  - Creates span hierarchy: root build span → package spans
  - Thread-safe with `sync.RWMutex` for concurrent builds
  - Captures build metrics and GitHub Actions context
  - Graceful degradation (tracing failures don't break builds)
  - Memory-safe: cleans up package spans after completion

- **CLI Integration** (`cmd/build.go`):
  - `--otel-endpoint` - OTLP endpoint URL
  - `--otel-insecure` - Disable TLS for OTLP endpoint
  - `--trace-parent` - W3C traceparent header
  - `--trace-state` - W3C tracestate header
  - Environment variable support with proper precedence

- **Dev Container** (`.devcontainer/devcontainer.json`):
  - Added Docker-in-Docker feature for testing tracing with local Jaeger

### Span Attributes

**Root Span (`leeway.build`):**
- `leeway.version`, `leeway.workspace.root`, `leeway.target.package`, `leeway.target.version`
- `leeway.packages.total`, `leeway.packages.cached`, `leeway.packages.remote`, `leeway.packages.downloaded`, `leeway.packages.to_build`
- GitHub context: `github.workflow`, `github.run_id`, `github.job`, etc.

**Package Spans (`leeway.package`):**
- `leeway.package.name`, `leeway.package.type`, `leeway.package.version`, `leeway.package.builddir`
- `leeway.package.last_phase`, `leeway.package.duration_ms`
- `leeway.package.phase.{phase}.duration_ms` for each phase
- Test coverage metrics when available

**Note:** Build phases are captured as span attributes rather than nested spans. This provides simpler hierarchy and lower overhead. Nested phase spans may be added in future work if deeper timeline visualization is needed.

### Testing
- Tests for OTelReporter with in-memory exporters
- Tests for trace context parsing, formatting, validation
- Tests for build lifecycle, concurrent packages, parent context propagation
- Tests for error handling, memory cleanup, and edge cases
- All tests passing ✅

### Documentation
- Created `docs/observability.md` with architecture, configuration, examples
- Updated `README.md` with OpenTelemetry section
- Usage examples for Jaeger, GitHub Actions

## Usage Example

```bash
# Start Jaeger
docker run -d --name jaeger \
  -p 4318:4318 \
  -p 16686:16686 \
  jaegertracing/all-in-one:latest

# Build with tracing (insecure for local development)
export OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4318
export OTEL_EXPORTER_OTLP_INSECURE=true
go run . build :my-package

# Or using CLI flags
go run . build :my-package \
  --otel-endpoint=localhost:4318 \
  --otel-insecure

# View traces at http://localhost:16686
```

## Related

Closes [CLC-2106](https://linear.app/ona-team/issue/CLC-2106)

### CODE REVIEW

File: cmd/build.go [hunk 1]
```diff
@@ -209,9 +213,13 @@ func addBuildFlags(cmd *cobra.Command) {
...
 	cmd.Flags().Bool("fixed-build-dir", true, "Use a fixed build directory for each package, instead of based on the package version, to better utilize caches based on absolute paths (defaults to true)")
 	cmd.Flags().Bool("docker-export-to-cache", false, "Export Docker images to cache instead of pushing directly (enables SLSA L3 compliance)")
+	cmd.Flags().String("otel-endpoint", os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT"), "OpenTelemetry OTLP endpoint URL for tracing (defaults to $OTEL_EXPORTER_OTLP_ENDPOINT)")
+	cmd.Flags().Bool("otel-insecure", os.Getenv("OTEL_EXPORTER_OTLP_INSECURE") == "true", "Disable TLS for OTLP endpoint (for local development only, defaults to $OTEL_EXPORTER_OTLP_INSECURE)")
+	cmd.Flags().String("trace-parent", os.Getenv("TRACEPARENT"), "W3C Trace Context traceparent header for distributed tracing (defaults to $TRACEPARENT)")
+	cmd.Flags().String("trace-state", os.Getenv("TRACESTATE"), "W3C Trace Context tracestate header for distributed tracing (defaults to $TRACESTATE)")
 }
 
-func getBuildOpts(cmd *cobra.Command) ([]leeway.BuildOption, cache.LocalCache) {
+func getBuildOpts(cmd *cobra.Command) ([]leeway.BuildOption, cache.LocalCache, func()) {
```
[the remark this answers is not in the document]
Comment (leodido) [reply cmd/build.go]: Or, a self-documenting approach. Perhaps like this?

```suggestion
type CleanupFunc func()
func getBuildOpts(cmd *cobra.Command) ([]leeway.BuildOption, cache.LocalCache, CleanupFunc)
```

**Review by leodido — COMMENTED** [review]
The OpenTelemetry implementation looks good to me!

Left some suggestions and questions in line.

Furthermore, can you please rebase it on the latest `main` branch?

After that, and before merging, I'd suggest testing this against our monorepo to validate:
- OTel tracing works in production CI
- No regressions in build behavior
- Cache hit rates remain stable

**How to do that?** By releasing an RC release from your latest commit in this PR
(eg, `git tag -a v0.16.0-rc1 -m v0.16.0-rc1`).

This allows us to:

- Update our monorepo's Leeway version temporarily
- Run full CI pipeline with tracing enabled
- Validate in production before final merge
- Spare time regarding the Leeway version dance in the monorepo's devcontainer

**Review by leodido — APPROVED** [review]
LGTM
